src/main/kotlin/io/arrowkt/model/Account.kt

```diff
@@ -0,0 +1,151 @@
+
+sealed class Account(
+    open val no: String,
+    open val name: String,
+    open val dateOfOpen: Option<LocalDate>,
+    open val dateOfClose: Option<LocalDate>,
+    open val balance: Balance
+) {
+    companion object {
+        private fun validateAccountNo(no: String): ValidationResult<String> =
+            if (no.length < 5) "Account No has to be at least 5 characters long: found $no".invalidNel()
+            else no.validNel()
+
+        private fun validateOpenCloseDate(
+            od: LocalDate,
+            cd: Option<LocalDate>
+        ): ValidationResult<Pair<Option<LocalDate>, Option<LocalDate>>> =
+            cd.map { c ->
+                if (c.isBefore(od)) "Close date $c cannot be earlier than open date $od".invalidNel()
+                else Pair(od.some(), cd).validNel()
+            }.getOrElse { Pair(od.some(), cd).validNel() }
+
+        private fun validateRate(rate: BigDecimal): ValidationResult<BigDecimal> =
+            if (rate < BigDecimal.ZERO) "Interest rate $rate must be > 0".invalidNel()
+            else rate.validNel()
+
+        fun checkingAccount(
+            no: String,
+            name: String,
+            openDate: Option<LocalDate>,
+            closeDate: Option<LocalDate>,
+            balance: Balance
+        ): ErrorOr<CheckingAccount> = ValidationResult.applicative(NonEmptyList.semigroup<String>())
+            .map(
+                validateAccountNo(no),
+                validateOpenCloseDate(
+                    openDate.getOrElse { today() },
+                    closeDate
+                )
+            ) { (n: String, d: Pair<Option<LocalDate>, Option<LocalDate>>) ->
+                CheckingAccount(n, name, d.first, d.second, balance)
+            }.fix().toEither()
+
+        fun savingsAccount(
+            no: String,
+            name: String,
+            rate: BigDecimal,
+            openDate: Option<LocalDate>,
+            closeDate: Option<LocalDate>,
+            balance: Balance
+        ): ErrorOr<SavingsAccount> = ValidationResult.applicative(NonEmptyList.semigroup<String>())
+            .map(
+                validateAccountNo(no),
+                validateOpenCloseDate(
+                    openDate.getOrElse { today() },
+                    closeDate
+                ),
+                validateRate(rate)
+            ) { (n: String, d: Pair<Option<LocalDate>, Option<LocalDate>>, r: BigDecimal) ->
+                SavingsAccount(n, name, r, d.first, d.second, balance)
+            }.fix().toEither()
+
+        private fun validateAccountAlreadyClosed(a: Account): ValidationResult<Account> =
+            if (a.dateOfClose.isDefined()) "Account ${a.no} already closed".invalidNel()
+            else a.validNel()
+
+        private fun validateCloseDate(a: Account, cd: LocalDate): ValidationResult<LocalDate> =
+            if (cd.isBefore(a.dateOfOpen.orNull())) "Close date $cd cannot be earlier than open date ${a.dateOfOpen.orNull()}".invalidNel()
```

Comment:
The Scala version uses `dateOfOpen.get()` here :scream:

Reply:
Fix it then :D `a.dateOfOpen.fold({ cd.validNel() }, { ... })`